### ISSUE TITLE
Big endian register load/stores

### ISSUE THREAD

**Issue by lim1ts** [icicle-emu/icicle-emu#47]
I have been experimenting with big endian m68k architecture on Icicle-emu, and noticed something strange with the behavior. I did this by making changes to `builder.rs` and `debug.rs`. 

Builder.rs :
```rust
//builder.rs
Architecture::M68k => SpecConfig {
            path: "68000/data/languages/coldfire.slaspec",
            processor_spec_path: None,
            reg_pc: "PC",
            reg_sp: "SP",
            context: vec![generic::CTX],
            init_registers: vec![],
            temp_registers: vec![],
            calling_cov: CallingCovSpec {
                integers: vec![],
                stack_align: 4,
                stack_offset: 0,
            },
            on_boot: generic::on_boot,
        },
```
and debug.rs:
```rust
// debug.rs
        Architecture::M68k => &[
            "D0", "D1", "D2", "D3", "D4", "D5", "D6", "D7",    // Data Registers
            "A0", "A1", "A2", "A3", "A4", "A5", "A6", "SP",    // Address Registers
            "SR", "ZF", "NF", "VF",                             // Status Register and Zero Flag
            "PC",                                              // Program Counter
        ][..],
```
To test this, i wrote a very small harness and passed in my code snippet, and observed a strange behavior on some operations.
In this ghidra listing, `Param 1` is the `D0` register, which was initialized to 0x00000001. 

![p3](https://github.com/icicle-emu/icicle-emu/assets/8879716/719874d2-6cce-41b6-92d6-94d903e641cf)

**Expected behavior:** D0 should be set to 0x0000ffff
**Actual behavior:** D0 is set to 0xffff0001.
![p2](https://github.com/icicle-emu/icicle-emu/assets/8879716/ea13623a-710b-4008-b7e8-698d735629e9)
Seems like somehow the register stores/loads are done in little endian?

Other instructions are decoding correctly, and `cpu.arch.sleigh.big_endian` also shows that it is in big endian mode. 

The instruction (0x206e6) above generated the following snippet in cranelift's IL.
![p1](https://github.com/icicle-emu/icicle-emu/assets/8879716/800f6601-3be6-405d-b19e-f4593e6fe911)

I'm guessing that the `store_var` function in `icicle-jit/src/translate/mod.rs` needs to handle the endianess?

**Comment by rbran** [comment]
I found this bug very interesting, the problem seems to be how sub-registers are handled, more specifically how the `offset` value of a `VarNode` is interpreted.

In this case, we are modifying the Register `D0w`, AKA the bytes 2..4 or `D0`, because this is big-endian, it should modify the least significant bytes.

```
0           4
+-----------+
|    D0     |
+-----+-----+
| D0u | D0w |
+-----+-----+
0     2     4
```

D0: `VarNode { id: 10, offset: 0, size: 4}`}
D0w: `VarNode { id: 10, offset: 2, size: 2}`)

In this case, the sub-register offset 2 and size 2 of a register with offset 0 and size 4, in big-endian means that we are writing at offset 0, `4-2-2` => register size 4 minus sub-register size 2 minus offset 2.

I believe this could be fixed by having the sub-registers created, at compile time, by a endian-aware function, here: https://github.com/icicle-emu/icicle-emu/blob/a46f623f095cb06351cf494bd9838d26ba516d81/sleigh/sleigh-compile/src/lib.rs#L91

Buy maybe I'm not getting how the code generation should work, and the solution is just to have all read/writes (load/store) do a `from_be`/`from_le` instead.

I was able to fix this problem by using this diff:
```patch
diff --git a/sleigh/sleigh-compile/src/lib.rs b/sleigh/sleigh-compile/src/lib.rs
index de297f0..dd4083b 100644
--- a/sleigh/sleigh-compile/src/lib.rs
+++ b/sleigh/sleigh-compile/src/lib.rs
@@ -83,13 +83,21 @@ pub fn build_inner(mut parser: Parser, verbose: bool) -> Result<SleighData, Stri

         // Check if there is an existing mapping for this register, if so add it as an alias.
         if let Some(&(id, byte_offset)) = ctx.data.register_mapping.get(&reg.offset) {
-            let alias = RegisterAlias { offset: byte_offset as u16, size: reg.size, name };
-            ctx.data.registers[id as usize].aliases.push(alias);
+            let parent_reg = &mut ctx.data.registers[id as usize];
+            let offset = if symbols.endianness == ast::EndianKind::Big {
+                parent_reg.size as u16 - byte_offset as u16 - reg.size as u16
+            }
+            else {
+                byte_offset as u16
+            };
+
+            let alias = RegisterAlias { offset, size: reg.size, name };
+            parent_reg.aliases.push(alias);

             ctx.data.named_registers[idx] = NamedRegister {
                 name,
-                var: pcode::VarNode::new(id, 16).slice(byte_offset, reg.size.min(16) as u8),
-                offset: reg.offset,
+                var: pcode::VarNode::new(id, 16).slice(offset as u8, reg.size.min(16) as u8),
+                offset: offset as u32,
             };
             continue;
         }
```

**Comment by rbran** [comment]
I was able to fix this problem by using this diff, but this seems it break other architectures.

```patch
diff --git a/sleigh/sleigh-compile/src/lib.rs b/sleigh/sleigh-compile/src/lib.rs
index de297f0..dd4083b 100644
--- a/sleigh/sleigh-compile/src/lib.rs
+++ b/sleigh/sleigh-compile/src/lib.rs
@@ -83,13 +83,21 @@ pub fn build_inner(mut parser: Parser, verbose: bool) -> Result<SleighData, Stri

         // Check if there is an existing mapping for this register, if so add it as an alias.
         if let Some(&(id, byte_offset)) = ctx.data.register_mapping.get(&reg.offset) {
-            let alias = RegisterAlias { offset: byte_offset as u16, size: reg.size, name };
-            ctx.data.registers[id as usize].aliases.push(alias);
+            let parent_reg = &mut ctx.data.registers[id as usize];
+            let offset = if symbols.endianness == ast::EndianKind::Big {
+                parent_reg.size as u16 - byte_offset as u16 - reg.size as u16
+            }
+            else {
+                byte_offset as u16
+            };
+
+            let alias = RegisterAlias { offset, size: reg.size, name };
+            parent_reg.aliases.push(alias);

             ctx.data.named_registers[idx] = NamedRegister {
                 name,
-                var: pcode::VarNode::new(id, 16).slice(byte_offset, reg.size.min(16) as u8),
-                offset: reg.offset,
+                var: pcode::VarNode::new(id, 16).slice(offset as u8, reg.size.min(16) as u8),
+                offset: offset as u32,
             };
             continue;
         }
```

**Comment by mchesser** [comment]
Thanks for the bug report (and the analysis from @rbran).

I think at the time I thought that the SLEIGH spec defined register overlaps in LE format because I saw that there were different BE and LE definitions for some architectures (e.g., [mips](https://github.com/NationalSecurityAgency/ghidra/blob/master/Ghidra/Processors/MIPS/data/languages/mips.sinc#L196)).

Based on @rbran's patch I came up with #48 which seems to work without breaking any of my tests. The main difference is that there are two places where SLEIGH offsets are mapped to internal vars, once based on names but an additional time when attachments and dynamic register offsets are resolved. Another difference is that `NamedRegister.offset` still stores the original SLEIGH offset.

**Comment by mchesser** [comment]
Fixed in #48

**Comment by lim1ts** [comment]
![image](https://github.com/icicle-emu/icicle-emu/assets/8879716/cd772f37-2321-4a5e-a548-4b5cebeb894c)
Tested again with the patch.

**Comment by mchesser** [comment]
I missed your comment and merged things too quickly. I have some additional fixes here: #49

**Comment by lim1ts** [comment]
Tested. Seems to be working perfectly on my code base. Thanks for the speedy updates :)
